lib/Conversion/TorchToLinalg/TorchToLinalg.cpp

```diff
@@ -3723,6 +3723,142 @@ class ConvertAtenBroadcastToOp : public OpConversionPattern<AtenBroadcastToOp> {
+                            scalarVal);
+  }
+  if (tensorType.getRank() == 1)
+    return tensor;
+  SmallVector<ReassociationIndices> reassociation(1);
+  for (auto i : llvm::seq<int64_t>(0, tensorType.getRank()))
+    reassociation[0].push_back(i);
+  return b.create<linalg::TensorCollapseShapeOp>(loc, tensor, reassociation);
+}
+
+// Given `input`, `indices`, and `values` `index_put` is given by:
+// `input` is an ND tensor, `indices` is a list of tensors, `value` is an ND
+// tensor.
+// For ex: input is a 2d tensor, indices is a list containing two 1d tensors.
+//   for i in range(0, len(indices[0]))):
+//     for j in range(0, len(input.shape[0]))
+//      for k in range(0, len(input.shape[1]))
+//         if (indices[0][i] == j && indices[1][i] == k)
+//            input[j][k] = values[i];
+namespace {
+class ConvertAtenIndexPutOp : public OpConversionPattern<AtenIndexPutOp> {
+public:
+  using OpConversionPattern::OpConversionPattern;
+  LogicalResult
+  matchAndRewrite(AtenIndexPutOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+
+    if (failed(verifyLinalgCompatibleTypes(op, rewriter)))
+      return failure();
+
+    Location loc = op.getLoc();
+    Value input = adaptor.self();
+    Value indices = op.indices();
+    Value elemVal = adaptor.values();
+    auto inputType = input.getType().cast<RankedTensorType>();
+    unsigned inputRank = inputType.getRank();
+
+    SmallVector<Value> indicesTuple;
+    if (!getListConstructElements(indices, indicesTuple)) {
+      return rewriter.notifyMatchFailure(op,
+                                         "unimplemented: indices argument is "
+                                         "not constructed from ListConstruct.");
+    }
+    unsigned numOfIndices = indicesTuple.size();
+    SmallVector<Value> indicesVal =
+        getTypeConvertedValues(rewriter, loc, getTypeConverter(), indicesTuple);
+
+    // TODO: Support broadcasting along dim.
+    if (inputRank != numOfIndices)
+      return rewriter.notifyMatchFailure(
+          op, "invalid shape: Broadcasting value along dim is not supported");
+
+    // TODO: More comprehensive checks:-
+    //   1. Since broadcasting is not implemented the number of elements in each
+    //   indexes as well as the output tensor should be same.
+    //
+    //   2. Since broadcasting is not supported and no. of elements should be
+    //   same hence the shape of tensors should be equivalent, for example:
+    //   [x,y] and [x,y,1] or [x,y,1,1] is considered equivalent but [y,x] and
+    //   [x,y] are not considered equivalent.
+    //
+    //  3. There should be an assert that the index_values lies in the bounds of
+    //  the input tensor.
+
+    // The value tensor along with all the indices are collapsed to 1d tensor.
+    SmallVector<Value> collapsedIndices;
+    for (Value val : indicesVal)
+      collapsedIndices.push_back(collapseTo1dTensor(rewriter, loc, val));
```

Comment:
Also, I'm not sure this method of collapsing will generalize well in the case where the indices tensors don't have the same number of elements. You can achieve a rank broadcasting with the `linalg.generic` maps. For example, suppose we have the following tensors:

```
a: tensor<4, 7>
i: tensor<2, 3>
j: tensor<4, 2, 3>
new_vals: tensor<3>
```

Then you can have the following maps:

```
a: (a, b, c, d, e) -> (a, b)
i: (a, b, c, d, e) -> (d, e)
j: (a, b, c, d, e) -> (c, d, e)
new_vals: (a, b, c, d, e) -> (e)
```

In other words, the domain of the maps have `a.rank + max_rank(i, j, new_vals)` dimensions. The input tensor maps to the first set of dimensions, and all other tensors map to the rest of the dimensions.

Reply:
I have added Broadcasting as a TODO for now.